src/Store/VimStoreConnector.re

```diff
@@ -959,6 +991,43 @@ let start =
         state,
         synchronizeEditorEffect(state),
       )
+    | Command("terminal.normalMode") =>
+      let maybeBufferId =
+        state
+        |> Selectors.getActiveBuffer
+        |> Option.map(Oni_Core.Buffer.getId);
+
+      let maybeTerminalId =
+        maybeBufferId
+        |> Option.map(id =>
+             BufferRenderers.getById(id, state.bufferRenderers)
+           )
+        |> OptionEx.flatMap(
+             fun
+             | BufferRenderer.Terminal({id, _}) => Some(id)
+             | _ => None,
+           );
```

Comment:
This is a pattern I';ve seen a few times. Might be worth having a `BufferRenderer.ifTerminal` or something like that. Or maybe it's a sign of larger structural issues. No biggie though, but seems inconvenient if it happens frequently.

Reply:
Actually, the number of `option`s being juggled here does suggest there are larger structural issues. That certainly doesn't mean the solution is in any way obvious, but I find there's usually a better way to organize things when it gets this hairy. Although it also tends to require a change in the model to more accurately reflect the actual domain.

Reply:
> Actually, the number of options being juggled here does suggest there are larger structural issues.

Right, I thjink this is true - I think this goes back to our discussion about the relationship between the Buffer Renderer <-> Buffer (and <-> Editor).

I'm thinking that a better way to model this would be to have the 'renderer' associated with the 'editor'. An editor <-> renderer relationship is always 1-to-1, whereas a buffer to renderer might not be (ie - for something like markdown preview - we'd have two editors associated with the same markdown file - but one would use the default editor renderer, and one would use a markdown-preview-renderer). That could cut down one of these `option`s being juggled.

The other thorn is that the editor has a reference to a `bufferId`, that may or may not exist - so we still have two `option`s. Need to think more about having our domain model more accurately reflect this. Because the editor <-> buffer relationship is 1-to-many, it doesn't make sense for the editor to store the buffer (nor the buffer to the list of editors referencing it) - so we might need some higher-level model to accurately represent the domain.

But maybe if we do that first-level of renderer refactoring (move the renderer into some context the editor knows about it), and continue to refactor around this area, something will pop out...

Reply:
I'll leave this as-is for now - but there are definitely domain-level issues we need to improve here

Reply:
Yep, no problem.

I really should have recognized that this was the bugger/edit/renderer issue we've talked about. I like your thinking though.

> Because the editor <-> buffer relationship is 1-to-many, it doesn't make sense for the editor to store the buffer (nor the buffer to the list of editors referencing it)

Is it really 1-to-many? Or is it either 1-to-1 or 1-to-2, or some other finite set of possible states? If so, we can certainly model that. If not, we could separate the model we store from the "context" we use for rendering and updating, which could be derived from the stored model but maintain the invariant that an editor cannot exist without a buffer. Either way this will likely require large architectural changes to work well, so it's definitely a long game.

Reply:
> Is it really 1-to-many? Or is it either 1-to-1 or 1-to-2, or some other finite set of possible states?

Good question. The buffer <-> editor relationship is 1 to some finite limit - for example, for each split, we have a new editor pointing to potentially to the same buffer. If you create 5 splits for the same file, you'd have 5 editors for one buffer (in our current model).

A renderer <-> editor should be 1 <-> 1 at any given time, but could change or be toggled.

> which could be derived from the stored model but maintain the invariant that an editor cannot exist without a buffer

This would be great. We don't have this invariant modeled in our current state.

Thanks for the thoughts! Hopefully we can figure out a good plan for this to remove some of these warts.